src/phoenix/experimental/evals/functions/binary.py

```diff
@@ -45,14 +48,10 @@ def llm_eval_binary(
     """
 
     eval_template = normalize_template(template)
-
     prompts = map_template(dataframe, eval_template)
-
     responses = model.generate(prompts.to_list(), system_instruction)
-    rail_classes = set(rails)
-    return [
-        (rail_class if (rail_class := resp.strip()) in rail_classes else None) for resp in responses
-    ]
+    rails_set = set(rails)
+    return [_snap_to_rail(response, rails_set) for response in responses]
```

Comment:
Maybe we do snap to the rail as part of the pipeline so the fail failure is co-located with the LLM invocation?

Reply:
It could make sense to co-locate the snapping of the rail with the invocation of the LLM. At the moment, our LLM interface has a `generate` method that takes in a list of prompts instead of a single prompt.

https://github.com/Arize-ai/phoenix/blob/b07c63e0e32d0b441b7513007c104842023eff81/src/phoenix/experimental/evals/models/base.py#L92

It might make sense to change the interface so that `generate` only invokes the LLM once, and then we have the concept of a `Job` that is responsible for invoking the LLM multiple times and snapping to a rail.